### PR TITLE
Fixing BZ issues - 1781287 and 1779732

### PR DESCRIPTION
**Pull request by jackyalbo**
Fixes for the following issues, some of them not fully:
https://bugzilla.redhat.com/show_bug.cgi?id=1781287 - added handling of Invalid names to KubeCreateSkipExisting instead of just panicing 
https://bugzilla.redhat.com/show_bug.cgi?id=1779732 - added Delete support for pvstore as well as handling special cased of pool being used by buckets/account
https://bugzilla.redhat.com/show_bug.cgi?id=1777295 - added to the CLI special handling of pools being used by buckets/account - it will still hang on deleting directly from kubernetes api, 
we need to think what to do in order to provide better feedback for that kind of failed deletions

### CODE REVIEW

File: pkg/pvstore/pvstore.go [hunk 1]
```diff
@@ -101,6 +112,34 @@ func RunCreate(cmd *cobra.Command, args []string) {
...
+	}
+
+	nbClient := system.GetNBClient()
+
+	poolName := args[0]
+
+	err := nbClient.DeletePoolAPI(nb.DeletePoolParams{
+		Name: poolName,
+	})
+	if err != nil {
```
Comment (nimrod-becker) [inline pkg/pvstore/pvstore.go]: What if pv-pool doesn't exist?

Comment (jackyalbo) [reply pkg/pvstore/pvstore.go]: then it will panic... we prefer something else? I guess it is pretty bad if to happens

Comment (guymguym) [reply pkg/pvstore/pvstore.go]: Probably best to just continue if the pool doesn't exist, because it means the deletion is essentially successful

File: pkg/util/util.go [hunk 2]
```diff
@@ -207,6 +207,10 @@ func KubeCreateSkipExisting(obj runtime.Object) bool {
 		log.Printf("❌ Forbidden: %s %q: %s\n", gvk.Kind, objKey.Name, err)
 		return false
 	}
+	if errors.IsInvalid(err) {
```
Comment (guymguym) [inline pkg/util/util.go]: What does Invalid means - that the validations of the object failed, like missing a required field for example?

In which case did you get it? 
Is this is a static object that we create in the code? do we generally expect validation errors?

File: pkg/backingstore/reconciler.go [hunk 3]
```diff
@@ -303,7 +303,19 @@ func (r *Reconciler) ReconcileDeletion() error {
...
 		if err != nil {
-			return err
+			if rpcErr, isRPCErr := err.(*nb.RPCError); isRPCErr {
+				if rpcErr.RPCCode == "DEFAULT_RESOURCE" {
+					return util.NewPersistentError("DefaultResource",
+						fmt.Sprintf("DeletePoolAPI cannot complete because pool %q is an account default resource", r.PoolInfo.Name))
+				} else if rpcErr.RPCCode == "IN_USE" {
+					return util.NewPersistentError("ResourceInUse",
+						fmt.Sprintf("DeletePoolAPI cannot complete because pool %q has buckets attached", r.PoolInfo.Name))
+				} else {
```
Comment (guymguym) [inline pkg/backingstore/reconciler.go]: How about just removing these two `else` statements and just keep it as a fallback to `return err` in any other case?

File: pkg/util/util.go (hunk 2 above)
Comment (jackyalbo) [reply pkg/util/util.go]: we got it when the name of the bucket is illeagal: ```PANI[0004] ☠️  Panic Attack: [Invalid] ObjectBucketClaim.objectbucket.io "obc#1" is invalid: metadata.name: Invalid value: "obc#1"```

File: pkg/backingstore/backingstore.go [hunk 4]
```diff
@@ -420,7 +422,18 @@ func RunDelete(cmd *cobra.Command, args []string) {
...
+	poolinfo, err := nbClient.ReadPoolAPI(nb.ReadPoolParams{Name: backStore.Name})
+	if err != nil {
+		log.Fatalf("❌ Failed to read BackingStore info: %s", err)
+	} else if poolinfo.Undeletable == "IN_USE" {
+		log.Fatalf(`❌ Could not delete BackingStore %q in namespace %q as it is being used by one or more buckets`,
+			backStore.Name, backStore.Namespace)
+	} else if poolinfo.Undeletable == "DEFAULT_RESOURCE" {
+		log.Fatalf(`❌ Could not delete BackingStore %q in namespace %q as it is the default resource of one or more accounts`,
+			backStore.Name, backStore.Namespace)
+	} else if !util.KubeDelete(backStore) {
```
Comment (guymguym) [inline pkg/backingstore/backingstore.go]: I would avoid `else if` here - it's hard to follow the flow when the delete call is hidden inside a compound if statement.

File: pkg/backingstore/backingstore.go [hunk 5]
```diff
@@ -420,7 +422,18 @@ func RunDelete(cmd *cobra.Command, args []string) {
...
 	backStore.Spec = nbv1.BackingStoreSpec{}
 
-	if !util.KubeDelete(backStore) {
+	nbClient := system.GetNBClient()
+
+	poolinfo, err := nbClient.ReadPoolAPI(nb.ReadPoolParams{Name: backStore.Name})
+	if err != nil {
+		log.Fatalf("❌ Failed to read BackingStore info: %s", err)
+	} else if poolinfo.Undeletable == "IN_USE" {
+		log.Fatalf(`❌ Could not delete BackingStore %q in namespace %q as it is being used by one or more buckets`,
```
Comment (nimrod-becker) [inline pkg/backingstore/backingstore.go]:
```suggestion
		log.Fatalf(`❌ Could not delete BackingStore %q in namespace %q as it is being used by one or more buckets classes`,
```

File: pkg/backingstore/reconciler.go (hunk 3 above)
Comment (jackyalbo) [reply pkg/backingstore/reconciler.go]: changed to switch

**Comment by jackyalbo** [comment]
> Maybe worth adding the flows you fixed/added to test_cli_flow? This way we'd have regression testing on it

I will discuss with @liranmauda as the whole deletion part of the flow is marked out.

File: pkg/backingstore/backingstore.go (hunk 4 above)
Comment (jackyalbo) [reply pkg/backingstore/backingstore.go]: changed to switch

File: pkg/backingstore/reconciler.go (hunk 3 above)
Comment (jackyalbo) [reply pkg/backingstore/reconciler.go]: removed the else as recommended

File: pkg/pvstore/pvstore.go (hunk 1 above)
Comment (jackyalbo) [reply pkg/pvstore/pvstore.go]: added handling for doesn't exist - so all these cases will have same kind of responses on failures

File: pkg/backingstore/backingstore.go [hunk 6]
```diff
@@ -427,7 +429,28 @@ func RunDelete(cmd *cobra.Command, args []string) {
 	backStore.Namespace = options.Namespace
 	backStore.Spec = nbv1.BackingStoreSpec{}
 
-	if !util.KubeDelete(backStore) {
+	nbClient := system.GetNBClient()
+
+	continueDelete := true
+	poolinfo, err := nbClient.ReadPoolAPI(nb.ReadPoolParams{Name: backStore.Name})
+	if err != nil {
+		if rpcErr, isRPCErr := err.(*nb.RPCError); !isRPCErr || rpcErr.RPCCode != "NO_SUCH_POOL" {
```
Comment (guymguym) [inline pkg/backingstore/backingstore.go]: I think this statement is too long for a one liner...
You can break it:
```golang
	rpcErr, isRPCErr := err.(*nb.RPCError)
	if !isRPCErr || rpcErr.RPCCode != "NO_SUCH_POOL" {
```

File: pkg/backingstore/backingstore.go [hunk 7]
```diff
@@ -427,7 +429,28 @@ func RunDelete(cmd *cobra.Command, args []string) {
...
+			log.Fatalf(`❌ Failed to read BackingStore info: %s`, err)
+			continueDelete = false
```
Comment (guymguym) [inline pkg/backingstore/backingstore.go]: log.Fatalf() exits the process after logging, so no need for the code after it, so continueDelete is unneeded

File: pkg/backingstore/backingstore.go [hunk 8]
```diff
@@ -427,7 +429,28 @@ func RunDelete(cmd *cobra.Command, args []string) {
...
+		}
+	} else {
+		switch poolinfo.Undeletable {
```
Comment (guymguym) [inline pkg/backingstore/backingstore.go]: I think we need to fail in any case of `poolinfo.Undeletable != ""`

File: test/cli/test_cli_functions.sh [hunk 9]
```diff
@@ -26,7 +26,13 @@ function kuberun {
 function noobaa {
     local rc
     local run_with_timeout=false
+    local should_fail=false
     local silence=false
+    if [ "${1}" == "failure" ]
```
Comment (guymguym) [inline test/cli/test_cli_functions.sh]: I think this is becoming very confusing that we have an override of the noobaa command that has all these test options that do not exist in the CLI. At least I would rename it to test_noobaa or something.

File: pkg/options/options.go [hunk 10]
```diff
@@ -30,7 +30,7 @@ const (
 	// ContainerImageRepo is the repo of the default image url
 	ContainerImageRepo = "noobaa-core"
 	// ContainerImageTag is the tag of the default image url
-	ContainerImageTag = "5.3.0-master20191219"
+	ContainerImageTag = "5.3.0-master20191223"
```
Comment (guymguym) [inline pkg/options/options.go]: Just making sure - is this a good image for anyone using master?

File: pkg/backingstore/backingstore.go (hunk 8 above)
Comment (jackyalbo) [reply pkg/backingstore/backingstore.go]: also on deleting? are we sure it will always be permanent error? shouldn't we reconcile on that?

File: pkg/options/options.go (hunk 10 above)
Comment (jackyalbo) [reply pkg/options/options.go]: suppose to fix some annoying issues: https://github.com/noobaa/noobaa-core/pull/5851

File: test/cli/test_cli_functions.sh (hunk 9 above)
Comment (jackyalbo) [reply test/cli/test_cli_functions.sh]: changed to test_noobaa

File: pkg/backingstore/backingstore.go (hunk 8 above)
Comment (jackyalbo) [reply pkg/backingstore/backingstore.go]: failing on all undeletables as bucket_deleting usually suggests a real issue in the core

File: pkg/backingstore/backingstore.go (hunk 7 above)
Comment (jackyalbo) [reply pkg/backingstore/backingstore.go]: fixed

File: pkg/backingstore/backingstore.go (hunk 6 above)
Comment (jackyalbo) [reply pkg/backingstore/backingstore.go]: fixed
